diff --git a/Makefile b/Makefile
index becb95317..74c74eade 100644
--- a/Makefile
+++ b/Makefile
@@ -73,7 +73,7 @@ GITREVDATE=$(shell git log -n 1 --format="%cd" --date=format:%Y%m%d-%H%M%S)
 # Reduces build time and binary sizes considerably.
 # That's only needed if you use gdb or nm.
 # If you need that, build manually without these flags.
-GOFLAGS := "-ldflags=-s -w -X github.com/google/syzkaller/prog.GitRevision=$(REV) -X 'github.com/google/syzkaller/prog.gitRevisionDate=$(GITREVDATE)'"
+GOFLAGS := -gcflags "all=-N -l" -ldflags "-X github.com/google/syzkaller/prog.GitRevision=$(REV) -X 'github.com/google/syzkaller/prog.gitRevisionDate=$(GITREVDATE)'"
 
 GOHOSTFLAGS ?= $(GOFLAGS)
 GOTARGETFLAGS ?= $(GOFLAGS)
diff --git a/executor/MRs/empty.h b/executor/MRs/empty.h
new file mode 100644
index 000000000..e69de29bb
diff --git a/executor/embed.go b/executor/embed.go
index 9a8a83141..0a30b0bf0 100644
--- a/executor/embed.go
+++ b/executor/embed.go
@@ -13,7 +13,7 @@ import (
 	"regexp"
 )
 
-//go:embed common*.h kvm*.h android/*.h
+//go:embed common*.h kvm*.h android/*.h MRs/*.h
 var src embed.FS
 
 // CommonHeader contains all executor common headers used by pkg/csource to generate C reproducers.
@@ -24,13 +24,13 @@ var CommonHeader = func() []byte {
 		panic(err)
 	}
 	headers := make(map[string]bool)
-	for _, glob := range []string{"*.h", "android/*.h"} {
+	for _, glob := range []string{"*.h", "android/*.h", "MRs/*.h"} {
 		files, err := fs.Glob(src, glob)
 		if err != nil {
 			panic(err)
 		}
 		for _, file := range files {
-			if file == "common.h" || file == "common_ext_example.h" {
+			if file == "common.h" || file == "common_ext_example.h" || file == "MRs/empty.h" {
 				continue
 			}
 			headers[file] = true
@@ -42,13 +42,21 @@ var CommonHeader = func() []byte {
 	for {
 		relacedSomething := false
 		for file := range headers {
-			replace := []byte("#include \"" + path.Base(file) + "\"")
-			if !bytes.Contains(data, replace) {
-				replace = []byte("#include \"android/" + path.Base(file) + "\"")
-				if !bytes.Contains(data, replace) {
-					continue
+			replaceCandidates := [][]byte{
+				[]byte("#include \"" + path.Base(file) + "\""),
+				[]byte("#include \"android/" + path.Base(file) + "\""),
+				[]byte("#include \"MRs/" + path.Base(file) + "\""),
+			}
+			var replace []byte
+			for _, candidate := range replaceCandidates {
+				if bytes.Contains(data, candidate) {
+					replace = bytes.Clone(candidate)
+					break
 				}
 			}
+			if replace == nil {
+				continue
+			}
 			contents, err := src.ReadFile(file)
 			if err != nil {
 				panic(err)
diff --git a/pkg/rpcserver/rpcserver.go b/pkg/rpcserver/rpcserver.go
index 69fd9e179..d49c17538 100644
--- a/pkg/rpcserver/rpcserver.go
+++ b/pkg/rpcserver/rpcserver.go
@@ -9,6 +9,7 @@ import (
 	"errors"
 	"fmt"
 	"net/url"
+	"path/filepath"
 	"slices"
 	"sort"
 	"strings"
@@ -34,10 +35,11 @@ type Config struct {
 	vminfo.Config
 	Stats
 
-	VMArch string
-	VMType string
-	RPC    string
-	VMLess bool
+	VMArch   string
+	VMType   string
+	RPC      string
+	MrvioDir string
+	VMLess   bool
 	// Hash adjacent PCs to form fuzzing feedback signal (otherwise just use coverage PCs as signal).
 	UseCoverEdges bool
 	// Filter signal/comparisons against target kernel text/data ranges.
@@ -157,10 +159,11 @@ func New(cfg *mgrconfig.Config, mgr Manager, stats Stats, debug bool) (Server, e
 			Sandbox:    sandbox,
 			SandboxArg: cfg.SandboxArg,
 		},
-		Stats:  stats,
-		VMArch: cfg.TargetVMArch,
-		RPC:    cfg.RPC,
-		VMLess: cfg.VMLess,
+		Stats:    stats,
+		VMArch:   cfg.TargetVMArch,
+		RPC:      cfg.RPC,
+		MrvioDir: filepath.Join(cfg.Workdir, "mrvio"),
+		VMLess:   cfg.VMLess,
 		// gVisor coverage is not a trace, so producing edges won't work.
 		UseCoverEdges: cfg.Experimental.CoverEdges && cfg.Type != targets.GVisor,
 		// gVisor/Starnix are not Linux, so filtering against Linux ranges won't work.
@@ -449,6 +452,7 @@ func (serv *server) CreateInstance(id int, injectExec chan<- bool, updInfo dispa
 		debug:         serv.cfg.Debug,
 		debugTimeouts: serv.cfg.DebugTimeouts,
 		sysTarget:     serv.sysTarget,
+		mrvioDir:      serv.cfg.MrvioDir,
 		injectExec:    injectExec,
 		infoc:         make(chan chan []byte),
 		requests:      make(map[int64]*queue.Request),
diff --git a/pkg/rpcserver/runner.go b/pkg/rpcserver/runner.go
index 29e79bad5..bd4e4da08 100644
--- a/pkg/rpcserver/runner.go
+++ b/pkg/rpcserver/runner.go
@@ -8,13 +8,16 @@ import (
 	"errors"
 	"fmt"
 	"os"
+	"path/filepath"
 	"slices"
+	"strings"
 	"sync"
 	"time"
 
 	"github.com/google/syzkaller/pkg/cover"
 	"github.com/google/syzkaller/pkg/flatrpc"
 	"github.com/google/syzkaller/pkg/fuzzer/queue"
+	"github.com/google/syzkaller/pkg/hash"
 	"github.com/google/syzkaller/pkg/log"
 	"github.com/google/syzkaller/pkg/osutil"
 	"github.com/google/syzkaller/pkg/report"
@@ -22,6 +25,7 @@ import (
 	"github.com/google/syzkaller/prog"
 	"github.com/google/syzkaller/sys/targets"
 	"github.com/google/syzkaller/vm/dispatcher"
+	"github.com/google/uuid"
 )
 
 type Runner struct {
@@ -46,6 +50,7 @@ type Runner struct {
 	lastExec      *LastExecuting
 	updInfo       dispatcher.UpdateInfo
 	resultCh      chan error
+	mrvioDir      string
 
 	// The mutex protects all the fields below.
 	mu          sync.Mutex
@@ -275,6 +280,8 @@ func (runner *Runner) sendRequest(req *queue.Request) error {
 		panic(err)
 	}
 	runner.nextRequestID++
+	// Return the output so that we can capture the metamorphic relation violation
+	req.ReturnOutput = true
 	id := runner.nextRequestID
 	var flags flatrpc.RequestFlag
 	if req.ReturnOutput {
@@ -414,6 +421,25 @@ func (runner *Runner) handleExecResult(msg *flatrpc.ExecResult) error {
 			runner.convertCallInfo(msg.Info.Extra)
 		}
 	}
+	// Check if MR is violated, and if so, save the program to workdir/mrvio.
+	if len(msg.Output) > 0 && strings.Contains(string(msg.Output), "[SyzMeta]: MR is violated!") {
+		// Let's classify MR-violated test inputs.
+		// Extract MR-related pseudo-syscall sequence in the req and construct mrSeq.
+		mrSeq := ""
+		for i := range len(req.Prog.Calls) {
+			callname := req.Prog.CallName(i)
+			if strings.HasPrefix(callname, "syz_mr_") {
+				mrSeq += callname
+			}
+		}
+		sig := hash.Hash([]byte(mrSeq))
+		// Using hash value of mrSeq for classification
+		dir := filepath.Join(runner.mrvioDir, sig.String())
+		osutil.MkdirAll(dir)
+		id := uuid.New().String()[:8]
+		fn := filepath.Join(dir, id)
+		osutil.WriteFile(fn, req.Prog.Serialize())
+	}
 	status := queue.Success
 	var resErr error
 	if msg.Error != "" {
diff --git a/sys/targets/targets.go b/sys/targets/targets.go
index a623dbfb6..8bb69d4d9 100644
--- a/sys/targets/targets.go
+++ b/sys/targets/targets.go
@@ -578,7 +578,8 @@ var oses = map[string]osCommon{
 
 var (
 	commonCFlags = []string{
-		"-O2",
+		"-g",
+		"-O0",
 		"-pthread",
 		"-Wall",
 		"-Werror",
